### PR TITLE
Game Setup Phase: Deck Class

### LINKED ISSUE #16: Game setup phase: Deck class

### LINKED ISSUE #86: One turn of the game: Deck class
Fields

private final Deque<Card> deck
private final Random random

Methods

public Deck(Deque<Card>, Random random) 
public void shuffle()
public Card peekTop()
public Card removeTop()
public int size()
public Card peekBottom()
public List<Card> peekTopNCards(int n)
public Card removeBottom()
public void addCard(Card card)
public boolean isEmpty()

### PR DESCRIPTION
**Pull request by Sunw00ha**
added Deck.md for shuffle method BVA of deck class

### CODE REVIEW

**Review by jazlynho2028 — CHANGES_REQUESTED** [review]
Just one fix for handling exceptions. Otherwise looks good!

**Comment by jazlynho2028** [comment]
Just as a reminder if you haven't yet, please pull from main to get the updated build.gradle.kts file so you can check for test coverage and mutants. To isolate a class for pitest, temporarily change the following two lines in build.gradle.kts:

pitest {
targetClasses.set(listOf("domain.*", "ui.Controller"))
targetTests.set(listOf("domain.", "ui.*ControllerTests"))

to:

pitest {
targetClasses.set(listOf("domain.Deck"))
targetTests.set(listOf("domain.DeckTests"))

**Comment by jazlynho2028** [comment]
Just need to add the following three methods to Deck to allow one turn of game:

public void addCard(Card card) {
        deck.add(card);
    }

    public boolean isEmpty() {
        return deck.isEmpty();
    }

    List<Card> getCards() {
        return List.copyOf(deck);
    }

File: src/main/java/domain/Deck.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+            cards.set(randomIndex, currentCard);
+        }
+
+        this.deck.clear();
+        this.deck.addAll(cards);
+    }
+
+    public Card peekTop() {
+        if (this.deck.isEmpty()) {
+            throw new UnsupportedOperationException("Cannot peek top of empty deck.");
```
Comment (jazlynho2028) [inline src/main/java/domain/Deck.java]: Can you change this to an IllegalStateException?

File: src/main/java/domain/Deck.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        }
+
+        return this.deck.peekFirst();
+    }
+
+    public Card removeTop() {
+        if (this.deck.isEmpty()) {
+            throw new UnsupportedOperationException("Cannot remove top of empty deck.");
```
Comment (jazlynho2028) [inline src/main/java/domain/Deck.java]: Can you change this to an IllegalStateException?

File: src/main/java/domain/Deck.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        return this.deck.removeFirst();
+    }
+
+    public int size() {
+        return this.deck.size();
+    }
+
+    public Card peekBottom() {
+        if (this.deck.isEmpty()) {
+            throw new UnsupportedOperationException("Cannot peek bottom of empty deck.");
```
Comment (jazlynho2028) [inline src/main/java/domain/Deck.java]: Can you change this to an IllegalStateException?

File: src/main/java/domain/Deck.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        return this.deck.peekLast();
+    }
+
+    public List<Card> peekTopNCards(int n) {
+        if (n < 0) {
+            throw new IllegalArgumentException("Cannot peek a negative number of cards.");
+        }
+
+        if (n > this.deck.size()) {
+            throw new UnsupportedOperationException("Cannot peek more cards than are in the deck.");
```
Comment (jazlynho2028) [inline src/main/java/domain/Deck.java]: Can you change this to an IllegalStateException?

File: src/main/java/domain/Deck.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        }
+
+        List<Card> cards = new ArrayList<>(this.deck);
+        return new ArrayList<>(cards.subList(0, n));
+    }
+
+    public Card removeBottom() {
```
Comment (jazlynho2028) [inline src/main/java/domain/Deck.java:79]: Can you change this to an IllegalStateException?

File: src/main/java/domain/Deck.java (hunk 1 above)
Comment (jazlynho2028) [reply src/main/java/domain/Deck.java]: Change all error strings to the format "error.emptyDeck". Then add the line

error.emptyDeck=Cannot access card in empty deck

in english .properties file and do the same but in Spanish in the spanish file. See other classes for examples if needed.

Comment (Sunw00ha) [reply src/main/java/domain/Deck.java]: Done!

File: src/main/java/domain/Deck.java (hunk 2 above)
Comment (Sunw00ha) [reply src/main/java/domain/Deck.java]: Done!

File: src/main/java/domain/Deck.java (hunk 3 above)
Comment (Sunw00ha) [reply src/main/java/domain/Deck.java]: Done!

File: src/main/java/domain/Deck.java (hunk 4 above)
Comment (Sunw00ha) [reply src/main/java/domain/Deck.java]: Done!

File: src/main/java/domain/Deck.java (hunk 5 above)
Comment (Sunw00ha) [reply src/main/java/domain/Deck.java:79]: Done!

File: src/main/java/domain/Deck.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.List;
+import java.util.Random;
+
+public class Deck {
+    private static final String ERROR_EMPTY_DECK = "error.emptyDeck";
```
Comment (jazlynho2028) [inline src/main/java/domain/Deck.java]: Instead of having this constant for your error string, notice that you have multiple if blocks checking whether deck is empty, then throws this same exception. You should make a private helper method (no need to test) that checks this condition and throws to reduce duplicate code.

File: src/main/java/domain/Deck.java [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+            justification = "Deck intentionally stores injected mutable dependencies " +
+                    "to support controlled deck state and deterministic shuffle testing."
+    )
+    public Deck(Deque<Card> deck, Random random) {
+        this.deck = deck;
+        this.random = random;
+    }
+
+    public void shuffle() {
+        List<Card> cards = new ArrayList<>(this.deck);
```
Comment (jazlynho2028) [inline src/main/java/domain/Deck.java]: You don’t need the “this.” prefix in front of the field names except the constructor. Only the constructor needs it because the parameters are the same name as the fields.

File: src/main/java/domain/Deck.java [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+
+        return this.deck.peekLast();
+    }
+
+    public List<Card> peekTopNCards(int n) {
+        if (n < 0) {
+            throw new IllegalArgumentException("Cannot peek a negative number of cards.");
+        }
+
+        if (n > this.deck.size()) {
```
Comment (jazlynho2028) [inline src/main/java/domain/Deck.java]: I would make a different error string becuase the empty deck error could be misleading since the deck isn’t actually empty the moment the error is thrown.

File: src/main/java/domain/Deck.java (hunk 7 above)
Comment (jazlynho2028) [reply src/main/java/domain/Deck.java]: Please change this for all methods (cleaner and easier to understand)

File: src/main/java/domain/Deck.java [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+            cards.set(i, cards.get(randomIndex));
+            cards.set(randomIndex, currentCard);
+        }
+
+        this.deck.clear();
+        this.deck.addAll(cards);
+    }
+
+    public Card peekTop() {
+        if (this.deck.isEmpty()) {
```
Comment (jazlynho2028) [inline src/main/java/domain/Deck.java]: See R11

File: src/main/java/domain/Deck.java [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+            throw new IllegalStateException(ERROR_EMPTY_DECK);
+        }
+
+        return this.deck.peekFirst();
+    }
+
+    public Card removeTop() {
+        if (this.deck.isEmpty()) {
```
Comment (jazlynho2028) [inline src/main/java/domain/Deck.java]: See R11

File: src/main/java/domain/Deck.java [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+
+        return this.deck.removeFirst();
+    }
+
+    public int size() {
+        return this.deck.size();
+    }
+
+    public Card peekBottom() {
+        if (this.deck.isEmpty()) {
```
Comment (jazlynho2028) [inline src/main/java/domain/Deck.java]: See R11

File: src/main/java/domain/Deck.java [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+        if (n > this.deck.size()) {
+            throw new IllegalStateException(ERROR_EMPTY_DECK);
+        }
+
+        List<Card> cards = new ArrayList<>(this.deck);
+        return new ArrayList<>(cards.subList(0, n));
+    }
+
+    public Card removeBottom() {
+        if (this.deck.isEmpty()) {
```
Comment (jazlynho2028) [inline src/main/java/domain/Deck.java]: See R11

File: src/test/java/domain/DeckTests.java [hunk 13]
```diff
@@ -0,0 +1,757 @@
...
+        IllegalStateException exception = assertThrows(
+                IllegalStateException.class,
+                deck::peekBottom);
+
+        assertEquals(ERROR_EMPTY_DECK, exception.getMessage());
+        assertEquals(0, deck.size());
+    }
+
+    @Test
+    public void peekBottom_oneCardDeck_returnsBottomCard() {
```
Comment (jazlynho2028) [inline src/test/java/domain/DeckTests.java]: You could parameterize this test and the two below it.

File: src/test/java/domain/DeckTests.java [hunk 14]
```diff
@@ -0,0 +1,757 @@
...
+        IllegalStateException exception = assertThrows(
+                IllegalStateException.class,
+                deck::removeTop);
+
+        assertEquals(ERROR_EMPTY_DECK, exception.getMessage());
+        assertEquals(0, deck.size());
+    }
+
+    @Test
+    public void removeTop_oneCardDeck_returnsTopCard() {
```
Comment (jazlynho2028) [inline src/test/java/domain/DeckTests.java]: You could parameterize this test test and the two below it.

File: src/test/java/domain/DeckTests.java [hunk 15]
```diff
@@ -0,0 +1,757 @@
...
+
+import org.easymock.EasyMock;
+import org.junit.jupiter.api.Test;
+
+import java.util.List;
+import java.util.Random;
+import java.util.ArrayDeque;
+import java.util.Deque;
+
+import static org.junit.jupiter.api.Assertions.*;
```
Comment (jazlynho2028) [inline src/test/java/domain/DeckTests.java:17]: For my comments about parameterizing tests, it would be easier to execute if you had an expectedCards and actualCards assertEquals comparison. Then you don't need to check for size and contents separately.

File: src/test/java/domain/DeckTests.java [hunk 16]
```diff
@@ -0,0 +1,757 @@
...
+
+        deck.shuffle();
+
+        assertEquals(0, deck.size());
+
+        EasyMock.verify(mockRandom);
+    }
+
+    @Test
+    public void shuffle_oneCardDeck_deckOrderUnchanged() {
```
Comment (jazlynho2028) [inline src/test/java/domain/DeckTests.java:33]: You should use an expectedCards and actualCards comparison here. Also your shuffle check is the same for this method and the two below, so you can make a runShuffleTest(Deque<Card> originalCards, Random mockRandom, List<Card> expectedCards) private helper in the test file 

private void runShuffleTest(Deque<Card> originalCards, Random mockRandom, List<Card> expectedCards) {
    Deck deck = new Deck(originalCards, mockRandom);
    deck.shuffle();
    assertEquals(expectedCards.size(), deck.size());

    List<Card> actualCards = new ArrayList<>();
    while (deck.size() > 0) {
        actualCards.add(deck.removeTop());
    }
    assertEquals(expectedCards, actualCards);
}

Then you can use this helper in this test and the two below

File: src/test/java/domain/DeckTests.java [hunk 17]
```diff
@@ -0,0 +1,757 @@
...
+
+        IllegalStateException exception = assertThrows(
+                IllegalStateException.class,
+                deck::peekTop);
+
+        assertEquals(ERROR_EMPTY_DECK, exception.getMessage());
+        assertEquals(0, deck.size());
+    }
+
+    @Test
```
Comment (jazlynho2028) [inline src/test/java/domain/DeckTests.java]: You can parameterize this test and the two below it.

File: src/test/java/domain/DeckTests.java [hunk 18]
```diff
@@ -0,0 +1,757 @@
...
+        Deck deck = new Deck(cards, new Random());
+
+        List<Card> result = deck.peekTopNCards(0);
+
+        assertEquals(0, result.size());
+        assertEquals(0, deck.size());
+    }
+
+    @Test
+    public void peekTopNCards_emptyDeckAndPositiveCount_throwsIllegalStateException() {
```
Comment (jazlynho2028) [inline src/test/java/domain/DeckTests.java]: You can parameterize this test and the one below it and peekTopNCards_countGreaterThanDeckSize_throwsIllegalStateException()

File: src/test/java/domain/DeckTests.java [hunk 19]
```diff
@@ -0,0 +1,757 @@
...
+        assertSame(card1, result);
+        assertEquals(2, deck.size());
+        assertSame(card1, deck.peekTop());
+        assertSame(card1, deck.peekBottom());
+
+        EasyMock.verify(card1);
+    }
+
+    @Test
+    public void peekTopNCards_emptyDeckAndZeroCards_returnsEmptyList() {
```
Comment (jazlynho2028) [inline src/test/java/domain/DeckTests.java]: You can parameterize this test and peekTopNCards_zeroCount_returnsEmptyList

File: src/test/java/domain/DeckTests.java [hunk 20]
```diff
@@ -0,0 +1,757 @@
...
+
+        assertEquals(0, result.size());
+        assertEquals(TWO_CARDS, deck.size());
+        assertSame(card1, deck.peekTop());
+
+        EasyMock.verify(card1, card2);
+    }
+
+    @Test
+    public void peekTopNCards_oneCardCount_returnsTopCardOnly() {
```
Comment (jazlynho2028) [inline src/test/java/domain/DeckTests.java]: You can parameterize this test and the one below it and peekTopNCards_duplicateCards_returnsTopCardsInOrder

File: src/test/java/domain/DeckTests.java [hunk 21]
```diff
@@ -0,0 +1,757 @@
...
+        IllegalStateException exception = assertThrows(
+                IllegalStateException.class,
+                deck::removeBottom);
+
+        assertEquals(ERROR_EMPTY_DECK, exception.getMessage());
+        assertEquals(0, deck.size());
+    }
+
+    @Test
+    public void removeBottom_oneCardDeck_returnsBottomCard() {
```
Comment (jazlynho2028) [inline src/test/java/domain/DeckTests.java]: You can parameterize this test and the two below it

File: src/test/java/domain/DeckTests.java [hunk 22]
```diff
@@ -0,0 +1,757 @@
...
+
+        assertSame(card1, result);
+        assertEquals(ONE_CARD, deck.size());
+        assertSame(card1, deck.peekTop());
+
+        EasyMock.verify(card1);
+    }
+
+    @Test
+    public void addCard_emptyDeck_addsCardToDeck() {
```
Comment (jazlynho2028) [inline src/test/java/domain/DeckTests.java]: You can parameterize this test and the three below it

File: src/test/java/domain/DeckTests.java [hunk 23]
```diff
@@ -0,0 +1,757 @@
...
+    @Test
+    public void isEmpty_emptyDeck_returnsTrue() {
+        Deque<Card> cards = new ArrayDeque<>();
+        Deck deck = new Deck(cards, new Random());
+
+        assertTrue(deck.isEmpty());
+    }
+
+    @Test
+    public void isEmpty_oneCardDeck_returnsFalse() {
```
Comment (jazlynho2028) [inline src/test/java/domain/DeckTests.java]: You can parameterize this test and the one below it

File: src/test/java/domain/DeckTests.java [hunk 24]
```diff
@@ -0,0 +1,757 @@
...
+
+        Deck deck = new Deck(cards, new Random());
+
+        assertFalse(deck.isEmpty());
+
+        EasyMock.verify(card1, card2);
+    }
+
+    @Test
+    public void getCards_emptyDeck_returnsEmptyList() {
```
Comment (jazlynho2028) [inline src/test/java/domain/DeckTests.java]: you can parameterize all getCards tests

File: src/test/java/domain/DeckTests.java (hunk 15 above)
Comment (jazlynho2028) [reply src/test/java/domain/DeckTests.java:17]: See DJ's Game.java tests in wip-game-revision if you would like some good examples

**Review by jazlynho2028 — CHANGES_REQUESTED** [review]

File: src/main/java/domain/Deck.java (hunk 6 above)
Comment (Sunw00ha) [reply src/main/java/domain/Deck.java]: made a private checkNotEmpty() method

File: src/main/java/domain/Deck.java (hunk 7 above)
Comment (Sunw00ha) [reply src/main/java/domain/Deck.java]: Done!

File: src/main/java/domain/Deck.java (hunk 8 above)
Comment (Sunw00ha) [reply src/main/java/domain/Deck.java]: This is true. Made the change!

File: src/test/java/domain/DeckTests.java (hunk 13 above)
Comment (Sunw00ha) [reply src/test/java/domain/DeckTests.java]: Done!

File: src/test/java/domain/DeckTests.java (hunk 14 above)
Comment (Sunw00ha) [reply src/test/java/domain/DeckTests.java]: Done!

File: src/test/java/domain/DeckTests.java (hunk 15 above)
Comment (Sunw00ha) [reply src/test/java/domain/DeckTests.java:17]: Done!

File: src/test/java/domain/DeckTests.java (hunk 16 above)
Comment (Sunw00ha) [reply src/test/java/domain/DeckTests.java:33]: Done!

File: src/main/java/domain/Deck.java (hunk 6 above)
Comment (jazlynho2028) [reply src/main/java/domain/Deck.java]: No need to keep error string constants here because they change with the properties files, not just based on this variable

File: src/main/java/domain/Deck.java [hunk 25]
```diff
@@ -0,0 +1,99 @@
...
+
+    public Card peekBottom() {
+        checkNotEmpty();
+
+        return deck.peekLast();
+    }
+
+    public List<Card> peekTopNCards(int n) {
+        if (n < 0) {
+            throw new IllegalArgumentException("Cannot peek a negative number of cards.");
```
Comment (jazlynho2028) [inline src/main/java/domain/Deck.java]: add a "error.peekNegativeCards"

File: src/main/java/domain/Deck.java [hunk 26]
```diff
@@ -0,0 +1,99 @@
...
+        }
+
+        if (n > deck.size()) {
```
Comment (jazlynho2028) [inline src/main/java/domain/Deck.java:71]: replace with error.notEnoughCards

File: src/main/java/domain/Deck.java [hunk 27]
```diff
@@ -0,0 +1,99 @@
...
+        return deck.removeFirst();
+    }
+
+    public int size() {
+        return deck.size();
+    }
+
+    private void checkNotEmpty() {
+        if (deck.isEmpty()) {
+            throw new IllegalStateException(ERROR_EMPTY_DECK);
```
Comment (jazlynho2028) [inline src/main/java/domain/Deck.java]: replace with error.emptyDeck

File: src/main/java/domain/Deck.java [hunk 28]
```diff
@@ -0,0 +1,99 @@
...
+        return new ArrayList<>(cards.subList(0, n));
+    }
+
+    public Card removeBottom() {
+        checkNotEmpty();
+
+        return deck.removeLast();
+    }
+
+    public void addCard(Card card) {
```
Comment (jazlynho2028) [inline src/main/java/domain/Deck.java:85]: Sorry I didn't catch this earlier, but Game expects this method to addCard to the top of the Deck rather than at the bottom. For example, it makes more sense to discard a card on top of the pile.

File: src/main/java/domain/Deck.java [hunk 29]
```diff
@@ -0,0 +1,99 @@
+package domain;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.List;
+import java.util.Random;
+
+public class Deck {
```
Comment (jazlynho2028) [inline src/main/java/domain/Deck.java:10]: can you add a public void insertCardAt(Card card, int index) so that players can choose where to insert the exploding kitten after they defuse it?

File: src/test/java/domain/DeckTests.java [hunk 30]
```diff
@@ -0,0 +1,766 @@
...
+import java.util.Random;
+import java.util.stream.Stream;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+public class DeckTests {
+    private static final String ERROR_EMPTY_DECK = "error.emptyDeck";
+    private static final String ERROR_NOT_ENOUGH_CARDS = "error.notEnoughCards";
+
+    private static final int ONE_CARD = 1;
```
Comment (jazlynho2028) [inline src/test/java/domain/DeckTests.java]: The following GameConstants.java file in main/java/domain already defined these. You should use them rather than defining them here.

package domain;

public final class GameConstants {

	private GameConstants() { }

	public static final int MIN_PLAYERS = 2;
	public static final int MAX_PLAYERS = 4;

	public static final int STARTING_HAND_SIZE = 6;
	public static final int STARTING_PLAYER_INDEX = 0;

	public static final int NUM_DEFUSES = 5;

	public static final int ONE_CARD = 1;
	public static final int TWO_CARDS = 2;
	public static final int THREE_CARDS = 3;
	public static final int FOUR_CARDS = 4;
	public static final int FIVE_CARDS = 5;

}

File: src/test/java/domain/DeckTests.java [hunk 31]
```diff
@@ -0,0 +1,766 @@
...
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.List;
+import java.util.Random;
+import java.util.stream.Stream;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+public class DeckTests {
+    private static final String ERROR_EMPTY_DECK = "error.emptyDeck";
```
Comment (jazlynho2028) [inline src/test/java/domain/DeckTests.java]: No need to define these as constants because the strings are already keys themselves in the .properties files. Just use the keys directly

**Review by jazlynho2028 — CHANGES_REQUESTED** [review]

File: docs/bva/Deck.md [hunk 32]
```diff
@@ -0,0 +1,239 @@
...
+  - **Name of the test**: `getCards_called_returnsCopyInDeckOrder` with case `multiple different cards`
+  - **State of the system**: Deck has cards without duplicates: [card1, card2]
+  - **Expected output**: Returns [card1, card2] in the same order; deck is still [card1, card2]
+
+- **TC39: Get cards from multiple-card deck with duplicates** ( :white_check_mark: )
+  - **Name of the test**: `getCards_called_returnsCopyInDeckOrder` with case `multiple duplicate cards`
+  - **State of the system**: Deck has cards with duplicates: [card1, card1, card2]
+  - **Expected output**: Returns [card1, card1, card2] with both copies of card1 preserved in order; deck is still [card1, card1, card2]
+
+### Method under test: `insertCardAt(Card card, int index)`
```
Comment (jazlynho2028) [inline docs/bva/Deck.md:205]: Mark x to white_check_mark for all tests

**Review by jazlynho2028 — APPROVED** [review]
Looks good! Just make sure the bva is updated with white_check_mark for all test cases
